File: algorithms/[395]至少有 K 个重复字符的最长子串/solution.py
# topics = ["分治", "递归"]

import sys
from collections import Counter


class Solution:
    def longestSubstring(self, s: str, k: int) -> int:
        cnt = Counter(s)

        min_times, max_times = sys.maxsize, 0
        for times in cnt.values():
            min_times = min(min_times, times)
            max_times = max(max_times, times)

        # 递归终止条件
        if min_times >= k:
            return len(s)
        if max_times < k:
            return 0

        res = 0
        i = 0
        for idx, ch in enumerate(s):
            if cnt[ch] < k:
                res = max(res, self.longestSubstring(s[i:idx], k))
                i = idx + 1

        return max(res, self.longestSubstring(s[i:], k))

    def longestSubstring2(self, s: str, k: int) -> int:
        n = len(s)
        if n < k:
            return 0

        for c in set(s):
            if s.count(c) < k:
                return max(self.longestSubstring(t, k) for t in s.split(c))

        return n

File: algorithms/[395]至少有 K 个重复字符的最长子串/solution.ts
// topics = ["分治算法", "递归"]

function longestSubstring(s: string, k: number): number {
  const n = s.length;
  const cnt = new Map<string, number>();
  for (const ch of s) {
    cnt.set(ch, (cnt.get(ch) ?? 0) + 1);
  }

  let minTimes = Number.MAX_SAFE_INTEGER;
  let maxTimes = 0;
  for (const times of cnt.values()) {
    minTimes = Math.min(minTimes, times);
    maxTimes = Math.max(maxTimes, times);
  }

  if (minTimes >= k) {
    return n;
  }
  if (maxTimes < k) {
    return 0;
  }

  let res = 0;
  let idx = 0;
  for (let i = 0; i < n; i += 1) {
    const ch = s[i];
    if (cnt.get(ch)! < k) {
      res = Math.max(res, longestSubstring(s.slice(idx, i), k));
      idx = i + 1;
    }
  }

  return Math.max(res, longestSubstring(s.slice(idx, n), k));
}
